java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/RelationalSymbolicValue.java

```diff
@@ -132,22 +132,64 @@ public BooleanConstraint shouldNotInverse() {
-      List<ProgramState> nullConstraints = copiedConstraints.stream().map(ps -> ps.addConstraint(rightOp, ObjectConstraint.nullConstraint())).collect(Collectors.toList());
+    BooleanConstraint leftConstraint = initialProgramState.getConstraint(leftOp, BooleanConstraint.class);
+    BooleanConstraint rightConstraint = initialProgramState.getConstraint(rightOp, BooleanConstraint.class);
+    if (leftConstraint != null && rightConstraint == null && !shouldNotInverse().equals(booleanConstraint)) {
+      List<ProgramState> nullConstraints = copiedConstraints.stream()
+        .flatMap(ps -> rightOp.setConstraint(ps, ObjectConstraint.NULL).stream())
+        .map(ps -> ps.addConstraints(rightOp, ps.getConstraints(rightOp).remove(BooleanConstraint.class))
+      ).collect(Collectors.toList());
       return ImmutableList.<ProgramState>builder().addAll(copiedConstraints).addAll(nullConstraints).build();
     }
     return copiedConstraints;
   }
 
-  @Override
   protected List<ProgramState> copyConstraint(SymbolicValue from, SymbolicValue to, ProgramState programState, BooleanConstraint booleanConstraint) {
     ProgramState newState = programState;
     if (programState.canReach(from) || programState.canReach(to)) {
       newState = programState.addConstraint(this, booleanConstraint);
     }
-    return super.copyConstraint(from, to, newState, booleanConstraint);
+    return copyConstraintFromTo(from, to, newState, booleanConstraint);
+  }
+
+  private List<ProgramState> copyConstraintFromTo(SymbolicValue from, SymbolicValue to, ProgramState programState, BooleanConstraint booleanConstraint) {
+    List<ProgramState> states = new ArrayList<>();
+    states.add(programState);
+    PMap<Class<? extends Constraint>, Constraint> leftConstraints = programState.getConstraints(from);
+    if (leftConstraints != null) {
+      leftConstraints.forEach((d, c) -> {
+        List<ProgramState> newStates = new ArrayList<>();
+        for (ProgramState state : states) {
+          if(ObjectConstraint.class.equals(d)) {
+            if(((ObjectConstraint) c).isNull()) {
+              newStates.addAll(to.setConstraint(state, shouldNotInverse().equals(booleanConstraint) ? c : c.inverse()));
+            } else if(shouldNotInverse().equals(booleanConstraint)) {
+              newStates.addAll(to.setConstraint(state, c));
+            } else {
+              newStates.add(state);
+            }
+            continue;
+          } else if(BooleanConstraint.class.equals(d)) {
+            newStates.addAll(to.setConstraint(state, shouldNotInverse().equals(booleanConstraint) ? c : c.inverse()));
+            continue;
+          }
+          Constraint constraint = shouldNotInverse().equals(booleanConstraint) ? c : c.inverse();
+          if (constraint == null) {
+            PMap<Class<? extends Constraint>, Constraint> constraints = state.getConstraints(to);
+            if (constraints != null) {
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png 'Severity: CRITICAL') Refactor this code to not nest more than 3 if/for/while/switch/try statements. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS134)